### ISSUE TITLE
Type inference fails for class objects in dict literals

### ISSUE THREAD

**Issue by Wade-Roberts** [astral-sh/ty#3469]
# Bug: Type inference fails for class objects in dict literals with generics

## Description
The `ty` type checker incorrectly infers the type of dictionary literals containing class objects as values when used with generic type parameters. When a dict literal uses class objects (e.g., `{"id": MyClass}`), the type checker fails to recognize that the values match the expected `type[BaseType]` constraint, causing a spurious type error.

## Expected Behavior
A dictionary literal with class objects as values should match a function signature expecting `dict[str, type[BaseType]]` where `BaseType` is a TypeVar bound to a base class.

## Actual Behavior
The type checker reports an error like:
```
Expected dict[str, type[BaseType]], found dict[str, <class 'Dog'> | <class 'Cat'>]
```

This is contradictory because `<class 'Dog'>` and `<class 'Cat'>` ARE instances of `type[BaseType]` (since they're both subclasses of Base).

## Steps to Reproduce

```python
from typing import TypeVar

class Base:
    """Base class"""
    pass

class Dog(Base):
    """Dog subclass"""
    pass

class Cat(Base):
    """Cat subclass"""
    pass

# TypeVar bound to Base
BaseType = TypeVar("BaseType", bound=Base)

def register_handlers(handlers: dict[str, type[BaseType]]) -> None:
    """Register handler classes"""
    pass

# ❌ This fails with type error
register_handlers({
    "dog": Dog,
    "cat": Cat,
})
```

## Impact
- Developers cannot use the natural, readable syntax of dict literals with class objects
- Affects any generic function accepting `dict[str, type[T]]` parameters
- Particularly problematic for mapping patterns (e.g., SQLAlchemy field-to-model mappings)

### Version

0.0.36

**Comment by ibraheemdev** [comment]
Thanks for the report. Here's a minimized example:
```py
class X: ...

def f[T: X](_: list[type[T]]): ...

f([X]) # error: Expected `list[type[X]]`, found `list[<class 'X'>]`
```

I believe the issue here is that we cannot use `list[type[T]]` as type context when inferring the list literal as it contains an unspecialized type variable, meaning we don't end up widening `<class 'X'>` to `type[X]` internally, and so assignability fails because of invariance. Note that this works fine if `X` is marked `@final`, because in that case `type[X]` and `<class 'X'>` are identical types.

**Comment by ibraheemdev** [comment]
I guess the easy fix here is to widen class literals from `<class 'X'>` to `type[X]` when we see an annotation of `type[UnspecializedTypeVar]`. However, that would mean we get less precise results when the widening is unnecessary, e.g.,
```py
_: type[X] = reveal_type(X) # ideally: <class 'X'>
```

I think this is really just another case of https://github.com/astral-sh/ty/issues/2521 which would be fixed by all three of the proposed solutions to that issue (roughly in increasing order of complexity, as well as correctness?):
1. Re-inferring arguments after specialization.
2. Passing through the unspecialized `type[T]` to the list literal, and so inferring `list[type[T] | <class 'X'>]`, and then recursively specializing `T` after inferring the specialization for the call expression, simplifying to `list[type[X]]`.
3. Inferring a constraint set `∃T, T :> class 'X'` for the literal `X`.

**Comment by carljm** [comment]
This was a recent "regression" from https://github.com/astral-sh/ruff/pull/24793, but looking at that PR, I suspect that we were just getting away this in some cases before due to the fallback to `type[Unknown]`, and now we aren't.
